vm/systemSmartContracts/delegation.go

```diff
@@ -808,3 +1245,120 @@ func (d *delegation) IsContractEnabled() bool {
 func (d *delegation) IsInterfaceNil() bool {
 	return d == nil
 }
+
+func getTransferBackFromVMOutput(vmOutput *vmcommon.VMOutput) *big.Int {
+	transferBack := big.NewInt(0)
+	for _, outAcc := range vmOutput.OutputAccounts {
+		for _, outTransfer := range outAcc.OutputTransfers {
+			transferBack.Add(transferBack, outTransfer.Value)
+		}
+	}
+
+	return transferBack
+}
+
+func moveNodeFromList(sndList []*NodesData, dstList []*NodesData, key []byte) {
+	var toMoveNodeData *NodesData
+	for i, nodeData := range sndList {
+		if bytes.Equal(nodeData.BLSKey, key) {
+			copy(sndList[i:], sndList[i+1:])
+			toMoveNodeData = nodeData
+			break
+		}
+	}
+
+	dstList = append(dstList, toMoveNodeData)
+}
+
+func isSuccessReturnData(returnData []byte) bool {
+	if bytes.Equal(returnData, []byte{waiting}) {
+		return true
+	}
+	if bytes.Equal(returnData, []byte{ok}) {
+		return true
+	}
+	return false
+}
+
+func getSuccessAndUnSuccessKeys(returnData [][]byte, blsKeys [][]byte) ([][]byte, [][]byte) {
+	if len(returnData) == 0 {
+		return blsKeys, nil
+	}
+
+	unSuccessKeys := make([][]byte, 0, len(returnData)/2)
+	for i := 0; i < len(returnData); i += 2 {
+		if !isSuccessReturnData(returnData[i+1]) {
```

Comment:
why aren't we check also on the provided blsKeys? What happens if the returned data contains something that is not contained by the blsKeys slice? If the check against the blsKeys slice is implemented here, the next for loop might be removed.

Reply:
this comes from checking how the auction smart contract returns these values. only for the failed bls keys, or bls keys which got into the waiting list will return. for those of success, nothing.

Reply:
I would still add a check against the provided blsKeys here in order to not have this situation in an exceptional case in which something goes wrong elsewhere:
```
returnData = "b"
blsKeys = "a1", "a2", "a3"
-> success will have "a1", "a2", "a3"
-> unsuccess will have "b"
```

Reply:
ok. will add len check.